fickling/fickle.py

```diff
@@ -539,14 +546,37 @@ def insert_function_call_on_unpickled_object(
         function_name = fn_match[0]
 
         # Insert exec of the function definition in advance
-        self.append_python(function_definition, attr="exec", pop_result=True)
+        if compile_code:
+            # Compile the function definition
+            code_obj = compile(function_definition, "<string>", "exec")
+            bytecode = marshal.dumps(code_obj) # marshal is required to get literal bytes
+
+            # Instructions:
+            ## Add the compiled bytes to the stack, then unmarshal them back into a code object
+            self.append_python(bytecode, module="marshal", attr="loads")
+
+            ## Slide exec instructions under the function definition on the stack
+            self.insert(-1, Put(1))  # Put function def in memory
+            self.insert(-1, Pop())
+            if not isinstance(self[-1], Stop):
+                raise ValueError("Expected the last opcode to be STOP")
+            ### NOTE(boyan): this seems to work even without insert GLOBAL at the beginning
+            ### of the pickle, but see comment in 'insert_python'
+            self.insert(-1, Global.create("builtins", "exec"))
+            self.insert(-1, Mark())
+            self.insert(-1, Get.create(1)) # Put the function def back
+            self.insert(-1, Tuple())
+            self.insert(-1, Reduce())
+            self.insert(-1, Pop()) # Remove extraneous return value 
+        else:
+            self.append_python(function_definition, attr="exec", pop_result=True)
 
         # Eval the function name get the callable object
         # If we inject myfunc() this will return eval(myfunc) which is the myfunc callable object
         self.append_python(function_name, attr="eval")
```

Comment:
@Boyan-MILANOV I wasn't able to think of a way to get around a plaintext exposure of `function_name`, since we're pretty heavily dependent on the `eval(function_name)` call here. Please let me know if you have any ideas

Reply:
TODO: I might be able to actually just compile a standalone string, need to test this

Reply:
For this you could simply use a function name that's a single char like `def a()`, that will go unnoticed in the raw pickle file.

Reply:
That's right, especially because it's essentially a call to eval("a") without any argument names being mentioned